File: src/ax_eval/harnesses/mock.py
"""Mock harness — a fake agent runtime for keyless, networkless runs.

It synthesizes a world state directly from each task's oracle specs, so a run is
deterministic and scoreable without touching a real target. To make a realistic,
*damning-demo* matrix (agents succeed at different rates), a mock can be told to
``skip`` some tasks (does nothing -> oracles fail) or get them ``wrong`` (writes
an off value -> equals/contains fail while exists passes).
"""

from __future__ import annotations

from typing import Any, Iterable

from ..datatypes import Task, TargetPack
from .base import Harness


def _set_path(world: dict[str, Any], path: str, value: Any) -> None:
    node = world
    parts = path.split(".")
    for part in parts[:-1]:
        node = node.setdefault(part, {})
    node[parts[-1]] = value


class MockHarness(Harness):
    """Deterministic fake harness with a configurable competence profile."""

    requires_key = False

    def __init__(
        self,
        name: str = "mock",
        *,
        skip: Iterable[str] = (),
        wrong: Iterable[str] = (),
    ) -> None:
        self.name = name
        self.skip = set(skip)
        self.wrong = set(wrong)

    def attempt(
        self, task: Task, pack: TargetPack
    ) -> tuple[dict[str, Any], list[str]]:
        trace = [f"[{self.name}] received task {task.id!r}: {task.title}"]

        if task.id in self.skip:
            trace.append(f"[{self.name}] gave up — produced no changes")
            return {}, trace

        world: dict[str, Any] = {}
        wrong = task.id in self.wrong
        for oracle in task.oracles:
            if not oracle.path:
                continue
            if oracle.type == "exists":
                _set_path(world, oracle.path, "" if wrong else "<created>")
            elif oracle.type == "equals":
                _set_path(
                    world,
                    oracle.path,
                    "__incorrect__" if wrong else oracle.expected,
                )
            elif oracle.type == "contains":
                _set_path(
                    world,
                    oracle.path,
                    [] if wrong else [oracle.value],
                )
        verb = "made plausible-but-wrong changes" if wrong else "completed the task"
        trace.append(f"[{self.name}] {verb}; reported world: {world}")
        return world, trace

File: src/adapters/mock.ts
/**
 * Mock harness — a fake agent runtime for keyless, networkless runs.
 *
 * It synthesizes a world state directly from each task's oracle specs, so a run
 * is deterministic and scoreable without touching a real target. To make a
 * realistic "damning demo" matrix (agents succeed at different rates), a mock
 * can be told to `skip` some tasks (does nothing → oracles fail) or get them
 * `wrong` (writes an off value → equals/contains fail while exists passes).
 */
import { HarnessAdapter, type AttemptOutcome } from "./base.js";
import type { Task, TargetPack, World } from "../schemas.js";

function setPath(world: World, path: string, value: unknown): void {
  const parts = path.split(".");
  let node = world;
  for (const part of parts.slice(0, -1)) {
    if (typeof node[part] !== "object" || node[part] === null) node[part] = {};
    node = node[part] as World;
  }
  node[parts[parts.length - 1]!] = value;
}

export interface MockOptions {
  skip?: string[];
  wrong?: string[];
}

export class MockHarness extends HarnessAdapter {
  readonly name: string;
  readonly requiresKey = false;
  protected readonly skip: Set<string>;
  protected readonly wrong: Set<string>;

  constructor(name = "mock", opts: MockOptions = {}) {
    super();
    this.name = name;
    this.skip = new Set(opts.skip ?? []);
    this.wrong = new Set(opts.wrong ?? []);
  }

  async attempt(task: Task, _pack: TargetPack): Promise<AttemptOutcome> {
    const trace = [`[${this.name}] received task '${task.id}': ${task.title}`];

    if (this.skip.has(task.id)) {
      trace.push(`[${this.name}] gave up — produced no changes`);
      return { world: {}, trace };
    }

    const world: World = {};
    const wrong = this.wrong.has(task.id);
    for (const oracle of task.oracles) {
      if (!oracle.path) continue;
      if (oracle.type === "exists") {
        setPath(world, oracle.path, wrong ? "" : "<created>");
      } else if (oracle.type === "equals") {
        setPath(world, oracle.path, wrong ? "__incorrect__" : oracle.expected);
      } else if (oracle.type === "contains") {
        setPath(world, oracle.path, wrong ? [] : [oracle.value]);
      }
    }
    const verb = wrong ? "made plausible-but-wrong changes" : "completed the task";
    trace.push(`[${this.name}] ${verb}; reported world: ${JSON.stringify(world)}`);
    return { world, trace };
  }
}
